rest/admin-ui-ext/src/main/java/org/keycloak/admin/ui/rest/AvailableRoleMappingResource.java

```diff
@@ -176,8 +212,35 @@ public final List<ClientRole> listCompositeUserRoleMappings(@PathParam("id") Str
                     )
             )}
     )
-    public final List<ClientRole> listCompositeRoleMappings(@QueryParam("first") @DefaultValue("0") long first,
+    public final List<ClientRole> listAvailableRoleMappings(@PathParam("id") String id, @QueryParam("first") @DefaultValue("0") long first,
             @QueryParam("max") @DefaultValue("10") long max, @QueryParam("search") @DefaultValue("") String search) {
-        return this.mapping(o -> true, first, max, search);
+        if (this.auth.users().canManage() || !Profile.isFeatureEnabled(Profile.Feature.ADMIN_FINE_GRAINED_AUTHZ)) {
+            this.auth.users().requireManage();
+            return searchForClientRolesByExcludedIds(realm, search, first, max, Stream.of(id));
+        } else {
+            Set<String> roleIds = getRoleIdsWithPermissions(MAP_ROLE_COMPOSITE_SCOPE, MAP_ROLES_COMPOSITE_SCOPE);
+            roleIds.remove(id);
+            return searchForClientRolesByIds(realm, roleIds.stream(), search, first, max);
+        }
+    }
+
+    private Set<String> getRoleIdsWithPermissions(String roleResourceScope, String clientResourceScope) {
+        Set<String> roleIds = this.auth.roles().getRolesWithPermission(roleResourceScope);
+        Set<String> clientIds = this.auth.clients().getClientsWithPermission(clientResourceScope);
+        //TODO: this might be pretty inefficient, directly joining client IDs with role IDs in one query would be better
```

Comment:
Please describe if you think this something you want to implement as part of this PR, or if it is good enough / as good as before, and you want handle this in a future issue/PR. 

Either way, there shouldn't be a TODO in the code for this. 

If you choose to have another issue for that, you can even reference the issue here to point to a discussion there.

Reply:
Thanks. I'll keep it as it is for now. The current code also populates the caches which has is upsides. I would revisit this should it turn out to be an issue in practice.